timemachine/cpp/src/random_sampler.hpp

```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "curand.h"
+#include "device_buffer.hpp"
+
+namespace timemachine {
+
+// An RandomSampler for random sampling from probabilities (np.random.choice(X, replace=True p=probabilities))
+template <typename RealType> class RandomSampler {
+
+private:
+    const int N_; // Max number of values that can be sampled from
+    size_t temp_storage_bytes_;
+
+    DeviceBuffer<RealType> d_rand_;
+    DeviceBuffer<RealType> d_gumbel_;
+
+    std::unique_ptr<DeviceBuffer<char>> d_sort_storage_;
+    // Can't import cub here to get the definition of KeyValuePair, create a single raw buffer dynamically and use casts
```

Comment:
what's the error that you get?

Reply:
It tries to compile the cub code with `gcc` which triggers a bunch of errors about `blockIdx.x` and such don't exist.

Reply:
Resolved with https://github.com/proteneer/timemachine/commit/12d5302dbc42a52a23d2904909af4605e9da2f16